File: code/backend/app/schemas/quiz.py
from __future__ import annotations

from datetime import datetime
from typing import Literal
from uuid import UUID

from pydantic import BaseModel, Field


class QuizOption(BaseModel):
    key: str = Field(min_length=1, max_length=8)
    text: str = Field(min_length=1, max_length=500)


class QuizQuestionDraft(BaseModel):
    knowledge_point: str = Field(min_length=1, max_length=160)
    question_type: Literal["single_choice"] = "single_choice"
    content: str = Field(min_length=1, max_length=2000)
    options: list[QuizOption] = Field(min_length=2, max_length=6)
    answer: str = Field(min_length=1, max_length=8)
    analysis: str = Field(min_length=1, max_length=3000)
    difficulty: Literal["foundation", "standard", "challenge"] = "standard"


class QuizDraft(BaseModel):
    title: str = Field(min_length=1, max_length=255)
    questions: list[QuizQuestionDraft] = Field(min_length=1, max_length=30)


class QuizQuestionPublic(BaseModel):
    id: UUID
    knowledge_point: str
    question_type: str
    content: str
    options: list[QuizOption]
    difficulty: str
    order: int


class QuizResourcePublic(BaseModel):
    resource_id: UUID
    title: str
    subject: str = "未分类"
    knowledge_point: str
    difficulty: str
    file_name: str = ""
    download_url: str = ""
    questions: list[QuizQuestionPublic]


class QuizSubmitRequest(BaseModel):
    answers: dict[str, str] = Field(default_factory=dict)


class QuizQuestionResult(BaseModel):
    question_id: UUID
    selected_answer: str = ""
    correct_answer: str
    is_correct: bool
    analysis: str
    knowledge_point: str
    saved_to_wrong_book: bool = False


class QuizSubmitResponse(BaseModel):
    attempt_id: UUID
    total_questions: int
    correct_count: int
    score: float
    wrong_knowledge_points: list[str]
    results: list[QuizQuestionResult]
    profile_updated: bool = True
    learning_path_updated: bool = True
    recommendation_refresh: bool = True


class QuizAttemptSummary(BaseModel):
    attempt_id: UUID
    resource_id: UUID
    total_questions: int
    correct_count: int
    score: float
    wrong_knowledge_points: list[str]
    created_time: datetime


class QuizAttemptDetail(QuizSubmitResponse):
    resource_id: UUID
    created_time: datetime


class WrongQuestionFavoriteRequest(BaseModel):
    favorite: bool = True


class WrongQuestionPublic(BaseModel):
    id: UUID
    question: QuizQuestionPublic
    resource_id: UUID
    resource_title: str
    subject: str = "未分类"
    wrong_count: int
    mastered: bool
    created_time: datetime
    updated_time: datetime


class WrongQuestionBookResponse(BaseModel):
    items: list[WrongQuestionPublic]
    count: int


class WrongBookSubmitResponse(BaseModel):
    total_questions: int
    correct_count: int
    score: float
    wrong_knowledge_points: list[str]
    results: list[QuizQuestionResult]
    attempt_ids: list[UUID]
    profile_updated: bool = True

File: code/education/course/src/api/quiz.ts
import axios from 'axios';

export interface QuizOption {
  key: string;
  text: string;
}

export interface QuizQuestion {
  id: string;
  knowledge_point: string;
  question_type: string;
  content: string;
  options: QuizOption[];
  difficulty: string;
  order: number;
}

export interface QuizResource {
  resource_id: string;
  title: string;
  knowledge_point: string;
  difficulty: string;
  questions: QuizQuestion[];
}

export interface QuizQuestionResult {
  question_id: string;
  selected_answer: string;
  correct_answer: string;
  is_correct: boolean;
  analysis: string;
  knowledge_point: string;
  saved_to_wrong_book: boolean;
}

export interface QuizSubmitResult {
  attempt_id: string;
  total_questions: number;
  correct_count: number;
  score: number;
  wrong_knowledge_points: string[];
  results: QuizQuestionResult[];
  profile_updated: boolean;
  learning_path_updated: boolean;
  recommendation_refresh: boolean;
}

export interface QuizAttemptSummary {
  attempt_id: string;
  resource_id: string;
  total_questions: number;
  correct_count: number;
  score: number;
  wrong_knowledge_points: string[];
  created_time: string;
}

export interface QuizAttemptDetail extends QuizSubmitResult {
  resource_id: string;
  created_time: string;
}

export interface WrongQuestionItem {
  id: string;
  question: QuizQuestion;
  resource_id: string;
  resource_title: string;
  subject: string;
  wrong_count: number;
  mastered: boolean;
  created_time: string;
  updated_time: string;
}

export interface WrongQuestionBook {
  items: WrongQuestionItem[];
  count: number;
}

export interface WrongBookSubmitResult {
  total_questions: number;
  correct_count: number;
  score: number;
  wrong_knowledge_points: string[];
  results: QuizQuestionResult[];
  attempt_ids: string[];
  profile_updated: boolean;
}

export interface WrongBookPracticeRequest {
  subject?: string;
  question_ids?: string[];
  count?: number;
  difficulty?: 'foundation' | 'standard' | 'challenge';
}

export function getQuiz(resourceId: string) {
  return axios.get<QuizResource>(`/api/resource-hub/quizzes/${resourceId}`);
}

export function submitQuiz(resourceId: string, answers: Record<string, string>) {
  return axios.post<QuizSubmitResult>(`/api/resource-hub/quizzes/${resourceId}/submit`, {
    answers,
  });
}

export function getQuizAttempts(resourceId: string) {
  return axios.get<QuizAttemptSummary[]>(`/api/resource-hub/quizzes/${resourceId}/attempts`);
}

export function getQuizAttempt(attemptId: string) {
  return axios.get<QuizAttemptDetail>(`/api/resource-hub/quiz-attempts/${attemptId}`);
}

export function setWrongQuestionFavorite(questionId: string, favorite: boolean) {
  return axios.put<{ question_id: string; favorite: boolean; wrong_count: number }>(
    `/api/resource-hub/wrong-book/${questionId}`,
    { favorite }
  );
}

export function getWrongQuestionBook() {
  return axios.get<WrongQuestionBook>('/api/resource-hub/wrong-book');
}

export function submitWrongQuestionBook(answers: Record<string, string>) {
  return axios.post<WrongBookSubmitResult>('/api/resource-hub/wrong-book/submit', { answers });
}

export function generateWrongQuestionPractice(payload: WrongBookPracticeRequest) {
  return axios.post<QuizResource>('/api/resource-hub/wrong-book/practice', payload);
}
